File: src/year2020/day12/day12.rs
//! [Day 12: Rain Risk](https://adventofcode.com/2020/day/12)

struct Instruction {
    action: char,
    n: i32,
}

struct Puzzle {
    instructions: Vec<Instruction>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            instructions: data
                .lines()
                .map(|line| Instruction {
                    action: line.chars().next().unwrap(),
                    n: line[1..].parse().unwrap(),
                })
                .collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> i32 {
        let mut x = 0;
        let mut y = 0;
        let mut d = 90;

        for Instruction { action, n } in &self.instructions {
            match action {
                'N' => y += n,
                'S' => y -= n,
                'E' => x += n,
                'W' => x -= n,
                'L' => d = (d - n).rem_euclid(360),
                'R' => d = (d + n).rem_euclid(360),
                'F' => match d {
                    0 => y += n,
                    90 => x += n,
                    180 => y -= n,
                    270 => x -= n,
                    _ => panic!(),
                },
                _ => panic!(),
            }
        }

        x.abs() + y.abs()
    }

    /// Solve part two.
    fn part2(&self) -> i32 {
        let (mut x, mut y) = (0, 0);
        let (mut wx, mut wy) = (10, 1);

        for Instruction { action, n } in &self.instructions {
            match action {
                'N' => wy += n,
                'S' => wy -= n,
                'E' => wx += n,
                'W' => wx -= n,
                'L' => match n {
                    90 => (wx, wy) = (-wy, wx),
                    180 => (wx, wy) = (-wx, -wy),
                    270 => (wx, wy) = (wy, -wx),
                    _ => panic!(),
                },
                'R' => match n {
                    270 => (wx, wy) = (-wy, wx),
                    180 => (wx, wy) = (-wx, -wy),
                    90 => (wx, wy) = (wy, -wx),
                    _ => panic!(),
                },
                'F' => (x, y) = (x + wx * n, y + wy * n),
                _ => panic!(),
            }
        }

        x.abs() + y.abs()
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (i32, i32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test_part1() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 25);
    }

    #[test]
    fn test_part2() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 286);
    }
}

File: src/year2020/day12/day12.py
#!/usr/bin/env python3
# [Day 12: Rain Risk](https://adventofcode.com/2020/day/12)

import atexit
import sys
import time
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


lines = data.splitlines()

# part 1

x, y, d = 0, 0, 90
for line in lines:
    c = line[0]
    n = int(line[1:])

    match c:
        case "N":
            y += n
        case "S":
            y -= n
        case "W":
            x -= n
        case "E":
            x += n
        case "L":
            d = (d - n) % 360
        case "R":
            d = (d + n) % 360
        case "F":
            match d:
                case 0:
                    y += n
                case 90:
                    x += n
                case 180:
                    y -= n
                case 270:
                    x -= n
                case _:
                    raise ValueError(f"bad direction {d}")
        case _:
            raise ValueError(f"bad action {line}")

    if verbose:
        print(f"{line:<5}  ship {x:5} {y:5}  direction {d:3}")

print(abs(x) + abs(y))


# part 2

x, y = 0, 0
wx, wy = 10, 1

for line in lines:
    c = line[0]
    n = int(line[1:])

    match c:
        case "N":
            wy += n
        case "S":
            wy -= n
        case "W":
            wx -= n
        case "E":
            wx += n
        case "L":
            match n:
                case 90:
                    wx, wy = -wy, wx
                case 180:
                    wx, wy = -wx, -wy
                case 270:
                    wx, wy = wy, -wx
                case _:
                    raise ValueError(f"bad angle {line}")

        case "R":
            match n:
                case 270:
                    wx, wy = -wy, wx
                case 180:
                    wx, wy = -wx, -wy
                case 90:
                    wx, wy = wy, -wx
                case _:
                    raise ValueError(f"bad angle {line}")
        case "F":
            x += wx * n
            y += wy * n
        case _:
            raise ValueError(f"bad action {c}")

    if verbose:
        print(f"{line:<5}  ship {x:5} {y:5}  waypoint {wx:5} {wy:5}")

print(abs(x) + abs(y))
